iceoryx_hoofs/source/posix_wrapper/shared_memory_object/shared_memory.cpp

```diff
@@ -33,73 +33,135 @@ namespace iox
+
+    if (!cxx::isValidFileName(m_name))
     {
-        std::cerr << "Shared memory name must start with a leading slash!" << std::endl;
-        m_isInitialized = false;
-        m_errorValue = SharedMemoryError::NAME_WITHOUT_LEADING_SLASH;
+        std::cerr << "Shared memory requires a valid file name (not path) as name and \"" << m_name
+                  << "\" is not a valid file name" << std::endl;
+        return cxx::error<SharedMemoryError>(SharedMemoryError::INVALID_FILE_NAME);
     }
 
-    if (m_isInitialized)
+    auto nameWithLeadingSlash = addLeadingSlash(m_name);
+
+    bool hasOwnership = (m_openMode == OpenMode::EXCLUSIVE_CREATE || m_openMode == OpenMode::PURGE_AND_CREATE
+                         || m_openMode == OpenMode::OPEN_OR_CREATE);
+    // the mask will be applied to the permissions, therefore we need to set it to 0
+    int sharedMemoryFileHandle = SharedMemory::INVALID_HANDLE;
+    mode_t umaskSaved = umask(0U);
     {
-        m_name = name;
-        m_isInitialized = open(accessMode, openMode, permissions, size);
+        cxx::GenericRAII umaskGuard([&] { umask(umaskSaved); });
+
+        if (m_openMode == OpenMode::PURGE_AND_CREATE)
+        {
+            IOX_DISCARD_RESULT(posixCall(iox_shm_unlink)(nameWithLeadingSlash.c_str())
+                                   .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                                   .ignoreErrnos(ENOENT)
+                                   .evaluate());
+        }
+
+        auto result =
+            posixCall(iox_shm_open)(
+                nameWithLeadingSlash.c_str(),
+                getOflagsFor(m_accessMode,
+                             (m_openMode == OpenMode::OPEN_OR_CREATE) ? OpenMode::EXCLUSIVE_CREATE : m_openMode),
+                static_cast<mode_t>(m_filePermissions))
+                .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                .suppressErrorMessagesForErrnos((m_openMode == OpenMode::OPEN_OR_CREATE) ? EEXIST : 0)
+                .evaluate();
+        if (result.has_error())
+        {
+            // if it was not possible to create the shm exclusively someone else has the
+            // ownership and we just try to open it
+            if (m_openMode == OpenMode::OPEN_OR_CREATE && result.get_error().errnum == EEXIST)
+            {
+                hasOwnership = false;
+                result = posixCall(iox_shm_open)(nameWithLeadingSlash.c_str(),
+                                                 getOflagsFor(m_accessMode, OpenMode::OPEN_EXISTING),
+                                                 static_cast<mode_t>(m_filePermissions))
+                             .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                             .evaluate();
+                if (!result.has_error())
+                {
+                    sharedMemoryFileHandle = result->value;
+                    return cxx::success<SharedMemory>(SharedMemory(m_name, sharedMemoryFileHandle, hasOwnership));
```

Comment:
`hasOwnership` is always false here. You could remove the line with `hasOwnership = false;` and use a `constexpr bool NO_OWNERSHIP {false};` here. Additionally `bool hasOwnership = ...` could be moved right before `if (hasOwnership)` where it is used for the first time.

What do you think?